File: samples/sample.go
package main

import (
	"fmt"
	"strings"
	"strconv"
	"encoding/json"
	"errors"
	"os"
)

// Single line comment
/* Multi-line
   comment */

// Constants
const (
	DefaultName = "John"
	DefaultAge  = 30
	Pi          = 3.14159
)

// Constants with iota (enumeration)
const (
	Red   = iota
	Green = iota
	Blue  = iota
)

// Type declarations
type Person struct {
	Name    string
	Age     int
	Email   string
	Address Address
}

type Address struct {
	Street string
	City   string
	State  string
	Zip    string
}

// Interface
type Reader interface {
	Read(p []byte) (n int, err error)
}

type Writer interface {
	Write(p []byte) (n int, err error)
}

// Embedded interface
type ReadWriter interface {
	Reader
	Writer
}

// Methods on types
func (p *Person) String() string {
	return fmt.Sprintf("%s (%d)", p.Name, p.Age)
}

func (p *Person) IsAdult() bool {
	return p.Age >= 18
}

// Numeric types and literals
var (
	int_var      int     = 42
	int8_var     int8    = 127
	int16_var    int16   = 32767
	int32_var    int32   = 2147483647
	int64_var    int64   = 9223372036854775807
	uint_var     uint    = 42
	uint8_var    uint8   = 255
	uint16_var   uint16  = 65535
	uint32_var   uint32  = 4294967295
	uint64_var   uint64  = 18446744073709551615
	uintptr_var  uintptr = 0xFF
	float32_var  float32 = 3.14
	float64_var  float64 = 3.14159
	complex64_v  complex64  = 1 + 2i
	complex128_v complex128 = 1 + 2i
)

// String and rune
var (
	str_var      string = "Hello, World!"
	rune_var     rune   = 'A'
	byte_var     byte   = 255
	raw_string   string = `raw\nstring`
)

// Boolean
var (
	bool_true  bool = true
	bool_false bool = false
)

// Arrays
var (
	arr_int    [5]int
	arr_string [3]string
	arr_init   [...]int{1, 2, 3, 4, 5}
)

// Slices
var (
	slice_int    []int
	slice_string []string
	slice_empty  []int
	slice_make   = make([]int, 5, 10)
)

// Maps
var (
	map_string_int = map[string]int{"a": 1, "b": 2}
	map_int_string = make(map[int]string)
	map_nested     = map[string]map[string]int{
		"group1": {"a": 1, "b": 2},
		"group2": {"c": 3, "d": 4},
	}
)

// Pointers
var (
	ptr_int    *int
	ptr_string *string
	ptr_person *Person
)

// Channels
var (
	ch_int   chan int
	ch_string chan string
	ch_send  chan<- int
	ch_recv  <-chan int
)

// Functions
func simpleFunction() {
	fmt.Println("Hello")
}

func functionWithArgs(name string, age int) {
	fmt.Printf("%s is %d years old\n", name, age)
}

func functionWithReturn(x int, y int) int {
	return x + y
}

func functionWithMultipleReturns(a int, b int) (int, error) {
	if b == 0 {
		return 0, errors.New("division by zero")
	}
	return a / b, nil
}

func functionWithNamedReturns(x int, y int) (sum int, product int) {
	sum = x + y
	product = x * y
	return
}

func functionWithVariadic(nums ...int) int {
	sum := 0
	for _, num := range nums {
		sum += num
	}
	return sum
}

// Defer
func deferExample() {
	defer fmt.Println("Last")
	fmt.Println("First")
	fmt.Println("Second")
}

// Panic and Recover
func panicExample() {
	defer func() {
		if r := recover(); r != nil {
			fmt.Println("Recovered:", r)
		}
	}()
	panic("Something went wrong")
}

// Anonymous function
func anonymousFunction() {
	f := func(x int) int {
		return x * 2
	}
	fmt.Println(f(5))
}

// Closure
func makeCounter() func() int {
	count := 0
	return func() int {
		count++
		return count
	}
}

// Goroutines and channels
func goroutineExample() {
	ch := make(chan string)
	go func() {
		ch <- "Hello from goroutine"
	}()
	msg := <-ch
	fmt.Println(msg)
}

// Select statement
func selectExample(ch1, ch2 chan string) {
	select {
	case msg1 := <-ch1:
		fmt.Println("Received:", msg1)
	case msg2 := <-ch2:
		fmt.Println("Received:", msg2)
	case <-time.After(time.Second):
		fmt.Println("Timeout")
	}
}

// Control flow
func controlFlow(x int) {
	// if-else
	if x > 0 {
		fmt.Println("Positive")
	} else if x < 0 {
		fmt.Println("Negative")
	} else {
		fmt.Println("Zero")
	}
	
	// if with short statement
	if y := x * 2; y > 10 {
		fmt.Println("Large")
	}
	
	// switch
	switch x {
	case 1:
		fmt.Println("One")
	case 2, 3:
		fmt.Println("Two or three")
		fallthrough
	default:
		fmt.Println("Other")
	}
	
	// switch with no condition
	switch {
	case x > 0:
		fmt.Println("Positive")
	case x < 0:
		fmt.Println("Negative")
	default:
		fmt.Println("Zero")
	}
	
	// for loop
	for i := 0; i < 5; i++ {
		fmt.Println(i)
	}
	
	// while loop
	i := 0
	for i < 5 {
		fmt.Println(i)
		i++
	}
	
	// infinite loop with break
	for {
		fmt.Println("Infinite")
		break
	}
	
	// for-range
	arr := []int{1, 2, 3, 4, 5}
	for i, v := range arr {
		fmt.Printf("Index: %d, Value: %d\n", i, v)
	}
	
	// for-range with strings
	for i, ch := range "Hello" {
		fmt.Printf("Index: %d, Char: %c\n", i, ch)
	}
	
	// for-range with maps
	m := map[string]int{"a": 1, "b": 2}
	for k, v := range m {
		fmt.Printf("Key: %s, Value: %d\n", k, v)
	}
	
	// break and continue
	for i := 0; i < 10; i++ {
		if i == 3 {
			continue
		}
		if i == 7 {
			break
		}
		fmt.Println(i)
	}
}

// Error handling
func errorHandling(filename string) error {
	_, err := os.Open(filename)
	if err != nil {
		return fmt.Errorf("failed to open file: %w", err)
	}
	return nil
}

// Operators
func operators() {
	a := 10
	b := 3
	
	// Arithmetic
	sum := a + b
	diff := a - b
	product := a * b
	quotient := a / b
	remainder := a % b
	power := a << 2  // Left shift
	rshift := a >> 1 // Right shift
	
	// Logical
	and := true && false
	or := true || false
	not := !true
	
	// Comparison
	eq := a == b
	neq := a != b
	lt := a < b
	gt := a > b
	lte := a <= b
	gte := a >= b
	
	// Bitwise
	bitand := a & b
	bitor := a | b
	bitxor := a ^ b
	bitnot := ^a
	lshift := a << 1
	rshift := a >> 1
	
	// Assignment
	a += 5
	b -= 2
	a *= 2
	b /= 2
	a %= 3
	
	// Ternary (using if-else)
	result := "positive"
	if a < 0 {
		result = "negative"
	}
}

// Comments with special syntax
// TODO: Implement this feature
// FIXME: Fix the bug
// NOTE: Important information
// HACK: Quick workaround

func main() {
	fmt.Println("Hello, World!")
	
	person := Person{
		Name: "John Doe",
		Age:  30,
		Email: "john@example.com",
	}
	
	fmt.Println(person)
}

File: samples/sample.rs
// Rust sample demonstrating all token types and syntax

// Comments
// Single line comment
/// Documentation comment for items
/** Multi-line
    documentation comment */
/*! Module documentation */
// Regular multi-line
/* comment */

// Prelude and imports
use std::collections::{HashMap, HashSet};
use std::io::{self, Read, Write};
use std::fs::File;

// Re-export
pub use crate::module::Item;

// Attributes
#![allow(dead_code)]
#![crate_type = "lib"]

#[derive(Debug, Clone, Copy)]
#[repr(C)]
struct Point {
    x: i32,
    y: i32,
}

// Constants and statics
const CONSTANT: i32 = 42;
const PI: f64 = 3.14159;
static STATIC_VAR: &str = "static";
static mut MUTABLE_STATIC: i32 = 0;

// Numeric types and literals
let decimal_int: i32 = 42;
let negative: i32 = -100;
let hex: i32 = 0xFF;
let octal: i32 = 0o755;
let binary: i32 = 0b1010;
let float32: f32 = 3.14;
let float64: f64 = 3.14159;
let scientific: f64 = 1.5e-3;
let underscored: i64 = 1_000_000;
let byte: u8 = b'A';
let inf: f64 = f64::INFINITY;
let neg_inf: f64 = f64::NEG_INFINITY;
let nan: f64 = f64::NAN;

// Strings and characters
let string: String = String::from("owned string");
let str_literal: &str = "string slice";
let raw_string: &str = r#"raw string with "quotes""#;
let byte_string: &[u8] = b"byte string";
let char_literal: char = 'A';
let escaped_char: char = '\n';
let unicode_char: char = '\u{1F600}';
let multi_line = "line1
line2";

// String interpolation
let name = "World";
let greeting = format!("Hello, {}!", name);

// Booleans
let bool_true: bool = true;
let bool_false: bool = false;

// Arrays
let array: [i32; 5] = [1, 2, 3, 4, 5];
let array_init: [i32; 3] = [0; 3];
let array_inference = [1, 2, 3];

// Tuples
let tuple: (i32, f64, &str) = (42, 3.14, "text");
let unit_tuple: () = ();
let (a, b, c) = tuple; // Destructuring

// Slices
let slice: &[i32] = &[1, 2, 3, 4, 5];
let slice_mutable: &mut [i32] = &mut [1, 2, 3];

// Vectors
let vector: Vec<i32> = vec![1, 2, 3];
let vector_new = Vec::new();
let vector_with_capacity = Vec::with_capacity(10);

// HashMaps
let mut map: HashMap<String, i32> = HashMap::new();
map.insert("a".to_string(), 1);
map.insert("b".to_string(), 2);

// HashSets
let mut set: HashSet<i32> = HashSet::new();
set.insert(1);
set.insert(2);

// Options
let option_some: Option<i32> = Some(42);
let option_none: Option<i32> = None;
if let Some(val) = option_some {
    println!("{}", val);
}

// Results
let result_ok: Result<i32, String> = Ok(42);
let result_err: Result<i32, String> = Err("error".to_string());
match result_ok {
    Ok(val) => println!("{}", val),
    Err(e) => println!("Error: {}", e),
}

// Enumerations
enum Color {
    Red,
    Green,
    Blue,
}

enum Message {
    Quit,
    Move { x: i32, y: i32 },
    Write(String),
    ChangeColor(i32, i32, i32),
}

// Structs
struct Person {
    name: String,
    age: u32,
    email: String,
}

struct Unit;

struct Tuple(String, i32, bool);

// Implementations
impl Point {
    fn new(x: i32, y: i32) -> Point {
        Point { x, y }
    }
    
    fn x(&self) -> i32 {
        self.x
    }
    
    fn translate(&mut self, dx: i32, dy: i32) {
        self.x += dx;
        self.y += dy;
    }
}

impl Person {
    fn new(name: String, age: u32) -> Person {
        Person {
            name,
            age,
            email: String::new(),
        }
    }
    
    fn display(&self) {
        println!("{} is {} years old", self.name, self.age);
    }
}

// Traits
trait Animal {
    fn name(&self) -> &str;
    
    fn make_sound(&self) {
        println!("{} makes a sound", self.name());
    }
}

struct Dog {
    name: String,
}

impl Animal for Dog {
    fn name(&self) -> &str {
        &self.name
    }
    
    fn make_sound(&self) {
        println!("{} says woof!", self.name());
    }
}

// Generic types
struct Container<T> {
    value: T,
}

impl<T> Container<T> {
    fn new(value: T) -> Container<T> {
        Container { value }
    }
    
    fn get(&self) -> &T {
        &self.value
    }
}

fn generic_function<T: std::fmt::Debug>(value: T) {
    println!("{:?}", value);
}

// Trait bounds
fn print_animal<T: Animal>(animal: T) {
    println!("{}", animal.name());
}

// Lifetime annotations
fn first_word<'a>(s: &'a str) -> &'a str {
    let bytes = s.as_bytes();
    for (i, &item) in bytes.iter().enumerate() {
        if item == b' ' {
            return &s[0..i];
        }
    }
    &s[..]
}

fn borrow_two<'a, 'b>(x: &'a str, y: &'b str) -> &'a str {
    x
}

// Functions
fn simple_function() {
    println!("Hello");
}

fn function_with_args(x: i32, y: i32) -> i32 {
    x + y
}

fn function_with_multiple_returns() -> (i32, String) {
    (42, "text".to_string())
}

fn function_with_impl_trait(x: i32) -> impl Iterator<Item = i32> {
    (0..x).map(|i| i * 2)
}

// Closures
let add_one = |x| x + 1;
let add = |x, y| x + y;
let closure_with_move = move |x| x + value;

// Control flow
fn control_flow(x: i32) {
    // if-else
    if x > 0 {
        println!("Positive");
    } else if x < 0 {
        println!("Negative");
    } else {
        println!("Zero");
    }
    
    // if as expression
    let result = if x > 0 { "positive" } else { "non-positive" };
    
    // match
    match x {
        0 => println!("Zero"),
        1..=5 => println!("Between 1 and 5"),
        10 | 20 => println!("Ten or twenty"),
        _ => println!("Other"),
    }
    
    // match with destructuring
    let point = (0, 0);
    match point {
        (0, 0) => println!("Origin"),
        (x, 0) => println!("On x-axis at {}", x),
        (0, y) => println!("On y-axis at {}", y),
        (x, y) => println!("At ({}, {})", x, y),
    }
    
    // loop
    loop {
        println!("Infinite");
        break;
    }
    
    // for loop
    for i in 0..5 {
        if i == 3 {
            continue;
        }
        if i == 4 {
            break;
        }
        println!("{}", i);
    }
    
    // while loop
    let mut i = 0;
    while i < 5 {
        println!("{}", i);
        i += 1;
    }
}

// Error handling
fn result_example(filename: &str) -> Result<String, std::io::Error> {
    let mut file = File::open(filename)?;
    let mut contents = String::new();
    file.read_to_string(&mut contents)?;
    Ok(contents)
}

// Ownership and borrowing
fn ownership_example() {
    let s1 = String::from("hello");
    let s2 = s1; // Move
    // println!("{}", s1); // Error: s1 is no longer valid
    
    let s3 = String::from("world");
    let s4 = &s3; // Borrow
    println!("{}", s3); // Valid, s3 is still valid
    
    let mut s5 = String::from("mutable");
    let s6 = &mut s5; // Mutable borrow
    s6.push_str(" string");
}

// Pattern matching
fn pattern_matching(value: Option<i32>) {
    match value {
        Some(x) if x > 5 => println!("Greater than 5"),
        Some(x) => println!("Some: {}", x),
        None => println!("None"),
    }
}

// Operators
fn operators() {
    let a = 10;
    let b = 3;
    
    // Arithmetic
    let _sum = a + b;
    let _diff = a - b;
    let _product = a * b;
    let _quotient = a / b;
    let _remainder = a % b;
    
    // Logical
    let _and = true && false;
    let _or = true || false;
    let _not = !true;
    
    // Comparison
    let _eq = a == b;
    let _neq = a != b;
    let _lt = a < b;
    let _gt = a > b;
    let _lte = a <= b;
    let _gte = a >= b;
    
    // Bitwise
    let _band = a & b;
    let _bor = a | b;
    let _bxor = a ^ b;
    let _bnot = !a;
    let _shl = a << 1;
    let _shr = a >> 1;
}

// Comments with special syntax
// TODO: Implement this feature
// FIXME: Fix the bug
// NOTE: Important information
// HACK: Quick workaround

fn main() {
    println!("Hello, world!");
    
    let point = Point::new(0, 0);
    println!("{:?}", point);
}
